compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,263 @@
+                                    ValueRange values) {
+  // Replace all OpResults with the given values.
+  SmallVector<Value> replacements;
+  for (auto [opResult, replacement] :
+       llvm::zip_equal(op->getOpResults(), values)) {
+    // If this value is a vector type, it must be converted back to simd.
+    if (isa<VectorType>(replacement.getType())) {
+      auto oldResult = cast<VectorValue>(opResult);
+      // Create a toSIMD op to convert the value back to the simd.
+      rewriter.setInsertionPointAfterValue(oldResult);
+      Value toSIMD = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+          oldResult.getLoc(), oldResult.getType(), replacement);
+      // Clone the layout to the new value.
+      options.getAnalysis().cloneLayoutInformationToNewValue(oldResult, toSIMD);
+      // Add to replacements.
+      replacement = toSIMD;
+    }
+    replacements.push_back(replacement);
+  }
+
+  rewriter.replaceOp(op, replacements);
+}
+
+class DistributeConstants : public OpRewritePattern<arith::ConstantOp> {
+public:
+  DistributeConstants(MLIRContext *context, VectorLayoutAnalysis &analysis,
+                      VectorLayoutOptions &options, PatternBenefit benefit = 1)
+      : OpRewritePattern<arith::ConstantOp>(context, benefit),
+        analysis(analysis), options(options) {}
+
+  LogicalResult matchAndRewrite(arith::ConstantOp constantOp,
+                                PatternRewriter &rewriter) const override {
+    Value constantResult = constantOp.getResult();
+    if (!isa<VectorType>(constantResult.getType()))
+      return failure();
+    auto constant = cast<VectorValue>(constantResult);
+
+    // Only handle splat values for now.
+    auto attr = dyn_cast<DenseElementsAttr>(constantOp.getValue());
+    if (!attr)
+      return failure();
+    if (!attr.isSplat())
+      return failure();
+
+    // Replace the original op with the distributed op.
+    Type elementType = constant.getType().getElementType();
+    auto vectorType =
+        VectorType::get(options.getDistributedShape(constant), elementType);
+    replaceOpWithNewDistributedOp<arith::ConstantOp>(
+        options, rewriter, constantOp, vectorType,
+        SplatElementsAttr::get(vectorType, attr.getSplatValue<Attribute>()));
+    return success();
+  }
+
+private:
+  VectorLayoutAnalysis &analysis;
+  VectorLayoutOptions &options;
+};
+
+class DistributeElementwise
+    : public OpTraitRewritePattern<OpTrait::Elementwise> {
+public:
+  DistributeElementwise(MLIRContext *context, VectorLayoutAnalysis &analysis,
+                        VectorLayoutOptions &options,
+                        PatternBenefit benefit = 1)
+      : OpTraitRewritePattern<OpTrait::Elementwise>(context, benefit),
+        analysis(analysis), options(options) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!OpTrait::hasElementwiseMappableTraits(op))
+      return failure();
```

Comment:
What's the difference between this and matching on `OpTrait::Elementwise`? I'd naively think they do the same thing?

Reply:
Ok I can see that elementwise is only one of the traits and there's no way to combine multiple in the pattern definition. Disregard this comment.